influx db revoke crashes vault

<!-- Please reserve GitHub issues for bug reports and feature requests.

For questions, the best place to get answers is on our [mailing list](https://groups.google.com/forum/#!forum/vault-tool), as they will get more visibility from experienced users than the issue tracker.

Please note: We take Vault's security and our users' trust very seriously. If you believe you have found a security issue in Vault, please responsibly disclose by contacting us at security@hashicorp.com. Our PGP key is available at [our security page](https://www.hashicorp.com/security/).

-->

**Describe the bug**
I'm using vault with nomad to run influxdb job and provide credentials to various other jobs (telegraf, etc.) through vault influxdb secrets engine. When I stop the nomad job, it revokes the given lease, after which the vault crashes.

I'm not sure if this is vault or nomad bug, but either way, I find my vault instance restarted and sealed because of this. I have started to see this behaviour after updating vault from `1.1.0` to `1.1.2` and nomad from `0.8.7` to `0.9.1`.

**Vault Logs**
It looks like vault revokes the influxdb lease correctly ( `database/creds/monitor` ), then fails to revoke two postgres leases (one is the role that is no longer existing, another a role that has some table ownerships, so it can't be revoked), after which it crashes. From the stacktrace it seems like it has some issue with influxdb lease, even tough it logged a successful revoke.

```
vault[29570]: 2019-05-15T09:54:27.196+0200 [INFO]  expiration: revoked lease: lease_id=database/creds/monitor/7qXaVgsw4GMZ2vsNhyDUPORi
vault[29570]: 2019-05-15T09:54:27.201+0200 [ERROR] expiration: failed to revoke lease: lease_id=database/creds/leaser_admin/H3JlBA2Pr2J4VcnrpCxswgnK error="failed to revoke entry: resp: (*logical.Res
vault[29570]: 2019-05-15T09:54:27.201+0200 [ERROR] expiration: failed to revoke lease: lease_id=database/creds/kms_admin/pi9DKryO8S1JyWzwCI0FVepO error="failed to revoke entry: resp: (*logical.Respon
vault[29570]: panic: runtime error: invalid memory address or nil pointer dereference
vault[29570]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x13f6a37]
vault[29570]: goroutine 3692 [running]:
vault[29570]: github.com/hashicorp/vault/vendor/github.com/influxdata/influxdb/client/v2.(*Response).Error(0x0, 0x0, 0xc0010a31c0)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/vendor/github.com/influxdata/influxdb/client/v2/client.go:477 +0x37
vault[29570]: github.com/hashicorp/vault/plugins/database/influxdb.(*Influxdb).RevokeUser(0xc001581cc0, 0x2f1a100, 0xc001244de0, 0xc0018967e0, 0x6b, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/plugins/database/influxdb/influxdb.go:194 +0x50f
vault[29570]: github.com/hashicorp/vault/builtin/logical/database/dbplugin.(*DatabaseErrorSanitizerMiddleware).RevokeUser(0xc0001e9e00, 0x2f1a100, 0xc001244de0, 0xc0018967e0, 0x6b, 0x0, 0x0, 0x0, 0x0
vault[29570]:         /gopath/src/github.com/hashicorp/vault/builtin/logical/database/dbplugin/databasemiddleware.go:236 +0xa3
vault[29570]: github.com/hashicorp/vault/builtin/logical/database/dbplugin.(*databaseMetricsMiddleware).RevokeUser(0xc001581ce0, 0x2f1a100, 0xc001244de0, 0xc0018967e0, 0x6b, 0x0, 0x0, 0x0, 0x0, 0x0, 
vault[29570]:         /gopath/src/github.com/hashicorp/vault/builtin/logical/database/dbplugin/databasemiddleware.go:148 +0x1fc
vault[29570]: github.com/hashicorp/vault/builtin/logical/database.(*databaseBackend).secretCredsRevoke.func1(0x2f1a100, 0xc001244de0, 0xc000f52a00, 0xc00074efd0, 0x0, 0x0, 0x0)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/builtin/logical/database/secret_creds.go:139 +0x47b
vault[29570]: github.com/hashicorp/vault/logical/framework.(*Secret).HandleRevoke(0xc0001e9aa0, 0x2f1a100, 0xc001244de0, 0xc000f52a00, 0xc00078d738, 0x1, 0x28)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/logical/framework/secret.go:87 +0x94
vault[29570]: github.com/hashicorp/vault/logical/framework.(*Backend).handleRevokeRenew(0xc00075cdd0, 0x2f1a100, 0xc001244de0, 0xc000f52a00, 0x393761376be6417b, 0x3530632d37303862, 0xc0008c98a0)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/logical/framework/backend.go:415 +0x1da
vault[29570]: github.com/hashicorp/vault/logical/framework.(*Backend).HandleRequest(0xc00075cdd0, 0x2f1a100, 0xc001244de0, 0xc000f52a00, 0x0, 0x0, 0x0)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/logical/framework/backend.go:170 +0x845
vault[29570]: github.com/hashicorp/vault/vault.(*Router).routeCommon(0xc000795c80, 0x2f1a100, 0xc001244de0, 0xc000f52a00, 0x40c100, 0x0, 0x2af0000, 0x0, 0x0)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/vault/router.go:665 +0x926
vault[29570]: github.com/hashicorp/vault/vault.(*Router).Route(0xc000795c80, 0x2f1a100, 0xc001244de0, 0xc000f52a00, 0xc001244de0, 0xc0008c9d48, 0x1f89071)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/vault/router.go:465 +0x4e
vault[29570]: github.com/hashicorp/vault/vault.(*ExpirationManager).revokeEntry(0xc00078c480, 0x2f1a100, 0xc00123c0c0, 0xc0010f2640, 0x2f, 0xc0010f2640)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/vault/expiration.go:1363 +0x284
vault[29570]: github.com/hashicorp/vault/vault.(*ExpirationManager).revokeCommon(0xc00078c480, 0x2f1a100, 0xc00123c0c0, 0xc001570540, 0x2f, 0xbf2f0f89203a0000, 0x0, 0x0)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/vault/expiration.go:626 +0x4b3
vault[29570]: github.com/hashicorp/vault/vault.(*ExpirationManager).Revoke(0xc00078c480, 0x2f1a100, 0xc00123c0c0, 0xc001570540, 0x2f, 0x0, 0x0)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/vault/expiration.go:573 +0x11c
vault[29570]: github.com/hashicorp/vault/vault.expireLeaseStrategyRevoke(0x2f1a040, 0xc000724040, 0xc00078c480, 0xc0010f34a0)
vault[29570]:         /gopath/src/github.com/hashicorp/vault/vault/expiration.go:122 +0x2c4
vault[29570]: github.com/hashicorp/vault/vault.(*ExpirationManager).updatePendingInternal.func1()
vault[29570]:         /gopath/src/github.com/hashicorp/vault/vault/expiration.go:1323 +0x52
vault[29570]: created by time.goFunc
vault[29570]:         /goroot/src/time/sleep.go:172 +0x44
systemd[1]: vault.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
systemd[1]: vault.service: Failed with result 'exit-code'.
```

--- a/plugins/database/influxdb/connection_producer.go
+++ b/plugins/database/influxdb/connection_producer.go
@@ -249,7 +249,7 @@ func isUserAdmin(cli influx.Client, user string) (bool, error) {
 	if err != nil {
 		return false, err
 	}
-	if response.Error() != nil {
+	if response != nil && response.Error() != nil {
 		return false, response.Error()
 	}
 	for _, res := range response.Results {

--- a/plugins/database/influxdb/connection_producer.go
+++ b/plugins/database/influxdb/connection_producer.go
@@ -249,7 +249,10 @@ func isUserAdmin(cli influx.Client, user string) (bool, error) {
 	if err != nil {
 		return false, err
 	}
-	if response != nil && response.Error() != nil {
+	if response == nil {
+		return false, fmt.Errorf("empty response")
+	}
+	if response.Error() != nil {
 		return false, response.Error()
 	}
 	for _, res := range response.Results {

--- a/plugins/database/influxdb/influxdb.go
+++ b/plugins/database/influxdb/influxdb.go
@@ -129,21 +129,25 @@ func (i *Influxdb) CreateUser(ctx context.Context, statements dbplugin.Statement
 				"password": password,
 			}), "", "")
 			response, err := cli.Query(q)
-			if err != nil && response.Error() != nil {
-				for _, stmt := range rollbackIFQL {
-					for _, query := range strutil.ParseArbitraryStringSlice(stmt, ";") {
-						query = strings.TrimSpace(query)
+			if err != nil {
+				if response != nil && response.Error() != nil {
+					for _, stmt := range rollbackIFQL {
+						for _, query := range strutil.ParseArbitraryStringSlice(stmt, ";") {
+							query = strings.TrimSpace(query)
 
-						if len(query) == 0 {
-							continue
-						}
-						q := influx.NewQuery(dbutil.QueryHelper(query, map[string]string{
-							"username": username,
-						}), "", "")
+							if len(query) == 0 {
+								continue
+							}
+							q := influx.NewQuery(dbutil.QueryHelper(query, map[string]string{
+								"username": username,
+							}), "", "")
 
-						response, err := cli.Query(q)
-						if err != nil && response.Error() != nil {
-							return "", "", err
+							response, err := cli.Query(q)
+							if err != nil {
+								if response != nil && response.Error() != nil {
+									return "", "", err
+								}
+							}
 						}
 					}
 				}
@@ -190,7 +194,9 @@ func (i *Influxdb) RevokeUser(ctx context.Context, statements dbplugin.Statement
 			}), "", "")
 			response, err := cli.Query(q)
 			result = multierror.Append(result, err)
-			result = multierror.Append(result, response.Error())
+			if response != nil {
+				result = multierror.Append(result, response.Error())
+			}
 		}
 	}
 	return result.ErrorOrNil()
@@ -230,7 +236,9 @@ func (i *Influxdb) RotateRootCredentials(ctx context.Context, statements []strin
 			}), "", "")
 			response, err := cli.Query(q)
 			result = multierror.Append(result, err)
-			result = multierror.Append(result, response.Error())
+			if response != nil {
+				result = multierror.Append(result, response.Error())
+			}
 		}
 	}
 

--- a/plugins/database/influxdb/influxdb.go
+++ b/plugins/database/influxdb/influxdb.go
@@ -131,31 +131,38 @@ func (i *Influxdb) CreateUser(ctx context.Context, statements dbplugin.Statement
 			response, err := cli.Query(q)
 			if err != nil {
 				if response != nil && response.Error() != nil {
-					for _, stmt := range rollbackIFQL {
-						for _, query := range strutil.ParseArbitraryStringSlice(stmt, ";") {
-							query = strings.TrimSpace(query)
-
-							if len(query) == 0 {
-								continue
-							}
-							q := influx.NewQuery(dbutil.QueryHelper(query, map[string]string{
-								"username": username,
-							}), "", "")
-
-							response, err := cli.Query(q)
-							if err != nil {
-								if response != nil && response.Error() != nil {
-									return "", "", err
-								}
-							}
-						}
-					}
+					attemptRollback(cli, username, rollbackIFQL)
 				}
 				return "", "", err
 			}
 		}
 	}
 	return username, password, nil
+}
+
+// attemptRollback will attempt to roll back user creation if an error occurs in
+// CreateUser
+func attemptRollback(cli influx.Client, username string, rollbackStatements []string) error {
+	for _, stmt := range rollbackStatements {
+		for _, query := range strutil.ParseArbitraryStringSlice(stmt, ";") {
+			query = strings.TrimSpace(query)
+
+			if len(query) == 0 {
+				continue
+			}
+			q := influx.NewQuery(dbutil.QueryHelper(query, map[string]string{
+				"username": username,
+			}), "", "")
+
+			response, err := cli.Query(q)
+			if err != nil {
+				if response != nil && response.Error() != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
 }
 
 // RenewUser is not supported on Influxdb, so this is a no-op.

--- a/plugins/database/influxdb/influxdb_test.go
+++ b/plugins/database/influxdb/influxdb_test.go
@@ -193,6 +193,56 @@ func TestMyInfluxdb_RenewUser(t *testing.T) {
 	}
 }
 
+// TestInfluxdb_RevokeDeletedUser tests attempting to revoke a user that was
+// deleted externally. Guards against a panic, see
+// https://github.com/hashicorp/vault/issues/6734
+func TestInfluxdb_RevokeDeletedUser(t *testing.T) {
+	if os.Getenv("VAULT_ACC") == "" {
+		t.SkipNow()
+	}
+	cleanup, address, port := prepareInfluxdbTestContainer(t)
+
+	connectionDetails := map[string]interface{}{
+		"host":     address,
+		"port":     port,
+		"username": "influx-root",
+		"password": "influx-root",
+	}
+
+	db := new()
+	_, err := db.Init(context.Background(), connectionDetails, true)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	statements := dbplugin.Statements{
+		Creation: []string{testInfluxRole},
+	}
+
+	usernameConfig := dbplugin.UsernameConfig{
+		DisplayName: "test",
+		RoleName:    "test",
+	}
+
+	username, password, err := db.CreateUser(context.Background(), statements, usernameConfig, time.Now().Add(time.Minute))
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	if err = testCredsExist(t, address, port, username, password); err != nil {
+		t.Fatalf("Could not connect with new credentials: %s", err)
+	}
+
+	// call cleanup to remove database
+	cleanup()
+
+	// attempt to revoke the user after database is gone
+	err = db.RevokeUser(context.Background(), statements, username)
+	if err == nil {
+		t.Fatalf("Expected err, got nil")
+	}
+}
+
 func TestInfluxdb_RevokeUser(t *testing.T) {
 	if os.Getenv("VAULT_ACC") == "" {
 		t.SkipNow()
@@ -301,7 +351,7 @@ func testCredsExist(t testing.TB, address string, port int, username, password s
 	if err != nil {
 		return errwrap.Wrapf("error querying influxdb server: {{err}}", err)
 	}
-	if response.Error() != nil {
+	if response != nil && response.Error() != nil {
 		return errwrap.Wrapf("error using the correct influx database: {{err}}", response.Error())
 	}
 	return nil

